File: src/sensor_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32_multi_array.hpp"
#include "visualization_msgs/msg/marker_array.hpp"
#include <vector>
#include <random>

class SensorNode : public rclcpp::Node
{
public:
    SensorNode() : Node("sensor_node"), num_sensors_(8)
    {
        pressure_pub_ = this->create_publisher<std_msgs::msg::Float32MultiArray>("/lateral_line/pressures", 10);
        marker_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>("/lateral_line/markers", 10);
        
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&SensorNode::publishSensors, this));
    }

private:
    void publishSensors()
    {
        // Simulate lateral line pressure values across array
        auto pressure_msg = std_msgs::msg::Float32MultiArray();
        pressure_msg.data.resize(num_sensors_);
        
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<float> dist(0.1f, 1.0f);

        for (int i = 0; i < num_sensors_; ++i) {
            pressure_msg.data[i] = dist(gen);
        }
        pressure_pub_->publish(pressure_msg);

        // Publish RViz2 visualization markers
        visualization_msgs::msg::MarkerArray marker_array;
        for (int i = 0; i < num_sensors_; ++i) {
            visualization_msgs::msg::Marker marker;
            marker.header.frame_id = "base_link";
            marker.header.stamp = this->now();
            marker.ns = "lateral_line";
            marker.id = i;
            marker.type = visualization_msgs::msg::Marker::CYLINDER;
            marker.action = visualization_msgs::msg::Marker::ADD;
            marker.pose.position.x = 0.2 - (i * 0.05);
            marker.pose.position.y = 0.2;
            marker.pose.position.z = 0.0;
            marker.scale.x = 0.02;
            marker.scale.y = 0.02;
            marker.scale.z = pressure_msg.data[i];
            marker.color.r = 0.0;
            marker.color.g = 0.5;
            marker.color.b = 1.0;
            marker.color.a = 0.8;
            marker_array.markers.push_back(marker);
        }
        marker_pub_->publish(marker_array);
    }

    rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr pressure_pub_;
    rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    int num_sensors_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SensorNode>());
    rclcpp::shutdown();
    return 0;
}

File: lateral_line_auv/sensor_node.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32MultiArray
from visualization_msgs.msg import Marker, MarkerArray
import random

class SensorNode(Node):
    def __init__(self):
        super().__init__('lateral_line_sensor_node')
        self.publisher_ = self.create_publisher(Float32MultiArray, '/lateral_line/pressures', 10)
        self.marker_pub_ = self.create_publisher(MarkerArray, '/visualization_marker_array', 10)
        self.timer = self.create_timer(0.2, self.timer_callback)
        self.get_logger().info('Lateral Line Sensor Node with RViz Markers Initialized.')

    def timer_callback(self):
        pressures = [random.uniform(0.1, 1.0) for _ in range(8)]
        
        msg = Float32MultiArray()
        msg.data = pressures
        self.publisher_.publish(msg)
        
        marker_array = MarkerArray()
        for i, p in enumerate(pressures):
            marker = Marker()
            marker.header.frame_id = 'base_link'
            marker.header.stamp = self.get_clock().now().to_msg()
            marker.ns = 'lateral_line_sensors'
            marker.id = i
            marker.type = Marker.CYLINDER
            marker.action = Marker.ADD
            marker.pose.position.x = 0.5 - (i * 0.1)
            marker.pose.position.y = 0.2 * (i - 3.5)
            marker.pose.position.z = 0.0
            marker.scale.x = 0.05
            marker.scale.y = 0.05
            marker.scale.z = float(p) * 0.5
            marker.color.r = float(p)
            marker.color.g = 0.2
            marker.color.b = 1.0 - float(p)
            marker.color.a = 0.8
            marker_array.markers.append(marker)
            
        self.marker_pub_.publish(marker_array)
        self.get_logger().info(f'Published pressure array & markers, obstacle ~{pressures[0]:.2f}m')

def main(args=None):
    rclpy.init(args=args)
    node = SensorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()